keyserver/listener.go

```diff
@@ -0,0 +1,127 @@
+package keyserver
+
+import (
+	"encoding/json"
+	"io"
+	"log"
+	"net"
+	"time"
+
+	. "github.com/coniks-sys/coniks-go/protocol"
+)
+
+func (server *ConiksServer) listenForRequests(ln net.Listener, handler func(msg []byte) ([]byte, error)) {
+	go func() {
+		// this is the only way to return from this function
+		<-server.stop
+		ln.Close()
+	}()
+
+	for {
+		conn, err := ln.Accept()
+		if err != nil {
+			select {
+			case <-server.stop:
+				return
+			default:
+				log.Printf("accept client: %s", err)
+				continue
+			}
+		}
+		server.waitStop.Add(1)
+		go func() {
+			server.acceptClient(conn, handler)
+			server.waitStop.Done()
+		}()
+	}
+}
+
+func (server *ConiksServer) acceptClient(conn net.Conn, handler func(msg []byte) ([]byte, error)) {
+	defer conn.Close()
+	closed := make(chan struct{})
+	defer close(closed)
+	go func() {
+		select {
+		case <-server.stop:
+			conn.Close()
```

Comment:
Oh, hmm, yeah, I think we should be a little nicer and give the whole 5 seconds for this connection to finish before shutting down.

With the above suggestions,  you can remove all the references to <-server.stop from this method.